File: frontend/lib/controllers/auth_controller.dart
import '../services/auth_service.dart';
import 'package:shared_preferences/shared_preferences.dart';

class AuthController {
  final AuthService _authService = AuthService();

  Future<Map<String, dynamic>> login(String email, String password) async {
    try {
      final result = await _authService.login(email, password);

      if (result["success"] == true) {
        final prefs = await SharedPreferences.getInstance();
        final token = result["data"]["token"];

        await prefs.setString("token", token);
        print("TOKEN = $token");
        await prefs.setInt(
          "userId",
          result["data"]["id"],
        );
        print(result);

        return {
          "success": true,
          "message": "Đăng nhập thành công"
        };
      }

      return {
        "success": false,
        "message": result["message"] ?? "Sai email hoặc mật khẩu"
      };

    } catch (e) {
      print(e);
      return {
        "success": false,
        "message": "Đã xảy ra lỗi hệ thống."
      };
    }
  }

  Future<void> logout() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove("token");
  }

  Future<bool> register(String username, String email, String password) async {
    try {
      final result = await _authService.register(username, email, password);
      return result["success"] == true;
    } catch (e) {
      print(e);
      return false;
    }
  }
}

File: backend/src/controllers/auth.controller.js
const authService = require("../services/auth.service");
const {
    saveLog
} = require(
    "../services/auditLog.service"
);
const register = async (req, res) => {
    try {

        console.log("BODY:", req.body);

        const { username, email, password } = req.body;

        const result = await authService.register(
            username,
            email,
            password
        );
        await saveLog(
            result,
            "REGISTER",
            "Đăng ký tài khoản mới"
        );

        res.status(201).json({
            success: true,
            data: result
        });

    } catch (error) {

        console.error("REGISTER ERROR:");
        console.error(error);

        res.status(400).json({
            success: false,
            message: error.message,
            stack: error.stack
        });
    }
};
const login = async (req, res) => {
    try {

        const { email, password } = req.body;

        const result = await authService.login(
            email,
            password
        );
        await saveLog(
            result,
            "LOGIN",
            "Đăng nhập thành công"
        );


        res.json({
            success: true,
            data: result
        });

    } catch (error) {
           console.log("LOGIN ERROR:");
            console.log(error);
        res.status(400).json({
            success: false,
            message: error.message
        });
    }
};

module.exports = {
    register,
    login
};